application/cache/cache.go

```diff
@@ -0,0 +1,266 @@
+	for _, privateRequest := range privateRequestsMap {
+		visibleRequestsList = append(visibleRequestsList, privateRequest)
+	}
+	for _, publicRequest := range publicRequestsMap {
+		visibleRequestsList = append(visibleRequestsList, publicRequest)
+	}
+
+	return visibleRequestsList, nil
+}
+
+// Rebuild cache after a request is created
+// We cache non-finished public requests publicly with cache key "requests-allusers"
+// We cache non-finished private caches privately with cache key "requests-orgname-affiliated_"
+// We remove a finished request from its respective cache
+func CacheRebuildOnNewRequest(ctx context.Context, request models.Request) error {
+	organization := request.Organization
+	// add to public cache only
+	if request.Visibility == models.RequestVisibilityAll {
+		newEntryCreated, requestsMap, err := getOrCreateCacheEntryPublic(ctx)
+		if err != nil {
+			return errors.New("error in cache rebuild (creation): " + err.Error())
+		}
+		if !newEntryCreated {
+			if err := updateRequestPublicCache(ctx, request, requestsMap); err != nil {
+				return errors.New("error in cache rebuild (creation): " + err.Error())
+			}
+		}
+	} else {
+		// add to private cache only
+		newEntryCreated, requestsMap, err := getOrCreateCacheEntryPrivate(ctx, organization)
+		if err != nil {
+			return errors.New("error in cache rebuild: " + err.Error())
+		}
+		if !newEntryCreated {
+			if err := updateRequestPrivateCache(ctx, request, requestsMap); err != nil {
+				return errors.New("error in cache rebuild: " + err.Error())
+			}
+		}
+	}
+	return nil
+}
+
+// Rebuild cache after a request is updated
+// We cache non-finished public requests publicly with cache key "requests-allusers"
+// We cache non-finished private caches privately with cache key "requests-orgname-affiliated_"
+// We remove a finished request from its respective cache
+func CacheRebuildOnChangedRequest(ctx context.Context, request models.Request) error {
+	organization := request.Organization
+	// update private cache
+	newEntryCreated, requestsMap, err := getOrCreateCacheEntryPrivate(ctx, organization)
+	if err != nil {
+		return errors.New("error in cache rebuild: " + err.Error())
+	}
+	if !newEntryCreated {
+		if err := updateRequestPrivateCache(ctx, request, requestsMap); err != nil {
+			return errors.New("error in cache rebuild: " + err.Error())
+		}
+	}
+
+	// update public cache
+	newEntryCreated, requestsMap, err = getOrCreateCacheEntryPublic(ctx)
+	if err != nil {
+		return errors.New("error in cache rebuild: " + err.Error())
+	}
+	if !newEntryCreated {
+		if err := updateRequestPublicCache(ctx, request, requestsMap); err != nil {
+			return errors.New("error in cache rebuild: " + err.Error())
+		}
+	}
+
+	return nil
+}
+
+// Gets cache entry for organization, or creates one if none exists
+// Returns true if new cache entry was created
+func getOrCreateCacheEntryPrivate(ctx context.Context, organization models.Organization) (bool, map[string]api.RequestAbridged, error) {
```

Comment:
This function and the next or so similar, I'm tempted to look for a way to factor out some of the code into a third shared function.  But, it's probably not worth it at this point.

Reply:
At this point, I'll leave refactoring as a TODO comment, but agreed they are similar enough that this is worth exploring later.

Reply:
They do use different SQL queries/call different models functions to populate cache entries, so a refactor will require some thoughtfulness.